code/datums/observation/moved.dm

```diff
@@ -15,38 +15,40 @@ var/decl/observ/moved/moved_event = new()
 	name = "Moved"
 	expected_type = /atom/movable
 
-/decl/observ/moved/register(var/eventSource, var/datum/procOwner, var/proc_call)
+/decl/observ/moved/register(var/atom/movable/mover, var/datum/listener, var/proc_call)
 	. = ..()
-	var/atom/movable/child = eventSource
-	if(.)
-		var/atom/movable/parent = child.loc
-		while(istype(parent) && !moved_event.is_listening(parent, child))
-			moved_event.register(parent, child, /atom/movable/proc/recursive_move)
-			child = parent
-			parent = child.loc
+
+	// Listen to the parent if possible.
+	if(. && istype(mover.loc, expected_type))
+		register(mover.loc, mover, /atom/movable/proc/recursive_move)
+
+/decl/observ/moved/unregister(var/atom/movable/mover, var/datum/listener, var/proc_call)
+	. = ..()
+
+	// Stop listening to the parent if we aren't being listened to.
+	if(. && !has_listeners(mover))
+		unregister(mover.loc, mover, /atom/movable/proc/recursive_move)
+
+// Handles the triggering of movement events by the parent.
+/atom/movable/proc/recursive_move(var/atom/movable/am, var/old_loc, var/new_loc)
+	moved_event.raise_event(src, old_loc, new_loc)
+
 
 /********************
 * Movement Handling *
 ********************/
-/atom/movable/proc/move_to_destination(var/atom/movable/am, var/old_loc, var/new_loc)
```

Comment:
This helper-proc is being used by myself in the say/radio rewrite PR, and will also be used by the AI eye for mob tracking.